contracts/LockedNORI.sol

```diff
@@ -40,127 +63,339 @@ contract LockedNORI is ERC777Upgradeable, ERC20PresetMinterPauserUpgradeable, ER
+    uint256 startTime,
+    uint256 vestEndTime,
+    uint256 unlockEndTime,
+    uint256 cliff1Time,
+    uint256 cliff2Time,
+    uint256 vestCliff1Amount,
+    uint256 vestCliff2Amount,
+    uint256 unlockCliff1Amount,
+    uint256 unlockCliff2Amount
+  ) public whenNotPaused onlyRole(TOKEN_GRANTER_ROLE) {
+    bytes memory userData = abi.encode(
+        recipient,
+        startTime,
+        vestEndTime,
+        unlockEndTime,
+        cliff1Time,
+        cliff2Time,
+        vestCliff1Amount,
+        vestCliff2Amount,
+        unlockCliff1Amount,
+        unlockCliff2Amount);
+    _grantTo(
+        amount,
+        userData
+    );
   }
 
-  function totalGrantAmount(address account) public view returns (uint256) {
+  /**
+   * @dev revokeUnvestedTokens: Truncates a vesting grant.
+   *
+   * Transfers any unvested tokens in *from*'s grant to *to*
+   * and reduces the total grant size.
+   *
+   * No change is made to balances that have vested but not claimed
+   * whether locked or not.
+   */
+  function revokeUnvestedTokens(
+    uint256 atTime,
+    address from,
+    address to
+  ) external onlyRole(TOKEN_GRANTER_ROLE) whenNotPaused {
+    TokenGrant storage grant = grants[from];
+    require(grant.exists, "No grant exist");
+    uint256 vestedBalance = _vestedBalanceOf(atTime, from);
+    if (vestedBalance < grant.grantAmount) {
+      uint256 quantityRevoked = grant.grantAmount - vestedBalance;
+      grant.grantAmount = vestedBalance;
+      emit UnvestedTokensRevoked(atTime, from, quantityRevoked);
+      operatorSend(
+        from,
+        to,
+        quantityRevoked,
+        "",
+        ""
+      );
+    }
+  }
+
+  /**
+   * @dev vestedBalanceOf: Vested balance less any claimed amount
+   */
+  function quantityRevokedFrom(address account) public view returns (uint256) {
     TokenGrant storage grant = grants[account];
-    return grant.grantAmount;
+    return grant.originalAmount - grant.grantAmount;
   }
 
+  /**
+   * @dev vestedBalanceOf: Vested balance less any claimed amount
+   */
   function vestedBalanceOf(address account) public view returns (uint256) {
+    return _vestedBalanceOf(block.timestamp, account);
+  }
+
+  /**
+   * @dev vestedBalanceOf: Vested balance less any claimed amount
+   */
+  function _vestedBalanceOf(uint256 atTime, address account)
+    internal
+    view
+    returns (uint256)
+  {
     TokenGrant storage grant = grants[account];
     if (grant.exists) {
-      // Past the end date user can claim any remaining wrapped tokens
-      return
-        grant.vestingSchedule.availableAmount(block.timestamp) -
-        grant.claimedAmount;
+      if (grant.vestingSchedule.startTime > 0) {
+        // If any tokens have been revoked then the schedule (which doesn't get updated) could
+        // return more than the total grant amount.
+        return
+          MathUpgradeable.min(grant.grantAmount, grant.vestingSchedule.availableAmount(atTime)) - grant.claimedAmount;
```

Comment:
can you break down why we need to take the minimum of these two instead of just the latter?

Reply:
In order to change the behavior of the lockup to the way Paul wants it to work -- unlock continues at the original rate after a revocation, I changed the implementation to shrink only `grant.grantAmount` (leaves the schedules untouched). So if `vestedBalanceOf` was being called after a revocation but before the end date the schedule amount would be greater than the total grant amount.

All that said that behavior is only relevant for unlocking not vesting and I can see why it's confusing.  I'll update it to shrink both `grant.grantAmount` and `grant.vestingSchedule.amount` and remove this check.  It's more explicit that way.